marklogic-client-api/src/test/java/com/marklogic/client/test/dataservices/BulkOutputCallerTest.java

```diff
@@ -0,0 +1,113 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+import com.marklogic.client.query.DeleteQueryDefinition;
+import com.marklogic.client.query.QueryManager;
+
+import static org.junit.Assert.*;
+
+public class BulkOutputCallerTest {
+    static ObjectNode apiObj;
+    static String apiName = "bulkOutputCallerImpl.api";
+    static String scriptPath;
+    static String apiPath;
+    static JSONDocumentManager docMgr;
+    static int count = 5;
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        docMgr = IOTestUtil.db.newJSONDocumentManager();
+        apiObj = IOTestUtil.readApi(apiName);
+        scriptPath = IOTestUtil.getScriptPath(apiObj);
+        apiPath = IOTestUtil.getApiPath(scriptPath);
+        IOTestUtil.load(apiName, apiObj, scriptPath, apiPath);
+        writeDocuments(count);
+    }
+    @Test
+    public void bulkOutputCallerWithNullConsumer() {
+        OutputEndpoint loadEndpt = OutputEndpoint.on(IOTestUtil.db, new JacksonHandle(apiObj));
+        OutputEndpoint.BulkOutputCaller loader = loadEndpt.bulkCaller();
+
+        expectedException.expect(IllegalStateException.class);
+        expectedException.expect(new ThrowableMessageMatcher(new StringContains("Output consumer is null")));
+        loader.awaitCompletion();
+    }
+
+    @Test
+    public void bulkOutputCallerTest() throws Exception {
+        String endpointState = "{\"next\":"+1+"}";
+        String workUnit      = "{\"max\":"+6+"}";
+
+        OutputCallerImpl caller = new OutputCallerImpl(new JacksonHandle(apiObj));
+
+        Stream<InputStream> result = caller.streamCall(IOTestUtil.db, IOTestUtil.asInputStream(endpointState), caller.newSessionState(),
+                IOTestUtil.asInputStream(workUnit));
+
+        InputStream[] resultArray = result.toArray(size -> new InputStream[size]);
+        assertNotNull(resultArray);
+        assertTrue(resultArray.length == count);
+        List<String> list = new ArrayList<>();
+        for(int i=0;i<resultArray.length;i++) {
+            assertNotNull(resultArray[i]);
+            list.add(IOTestUtil.mapper.readValue(resultArray[i], ObjectNode.class).toString());
+        }
+
+        OutputEndpointImpl.BulkOutputCaller bulkCaller = OutputEndpoint.on(IOTestUtil.db, new JacksonHandle(apiObj)).bulkCaller();
+        bulkCaller.setEndpointState(new ByteArrayInputStream(endpointState.getBytes()));
+        bulkCaller.setWorkUnit(new ByteArrayInputStream(workUnit.getBytes()));
+        bulkCaller.forEachOutput(i-> {
+            try {
+                assertTrue(list.contains(IOTestUtil.mapper.readValue(i, ObjectNode.class).toString()));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        });
+
+       bulkCaller.awaitCompletion();
```

Comment:
Would it make sense to confirm that forEachOutput() executed the expected number of times?

That is, should the lambda increment a counter and an assertion on the counter follow completion?